Auth flow

8888 opened the pull request:
WIP

Outstanding work to do:
- actual session management. Tokens do not timeout
- authorization header for user/token
- while usernames are unique, dbInterface uses that assumption when it should be using an ID

8888 committed:
initial db creation

created postgres db and schema on a docker image. updated readme with all steps taken and setup instructions to get back to this point

8888 committed:
npm install pg

8888 committed:
initial creation of DbInterface class

these are helper methods to interact directly with the database using the pg library.

8888 committed:
create Auth class and registerUser method

8888 committed:
create method to for user to login

this also needed a DbInterface method to fetch the salt

8888 committed:
npm install express

8888 committed:
npm install body-parser

8888 committed:
create express server and register endpoint

8888 committed:
add login endpoint

8888 committed:
allow cors for express

also return username when logging in as I will probably want to use this

8888 committed:
create index.html and index.js

This is a simple login form. The script has logic to send the credentials to the server to auth the user.

8888 committed:
create endpoint to confirm a users token is active

8888 committed:
add private page

a user should only be able to access this page when logged in. If theyn are not, they will be redirected to index

8888 committed:
add registration functionality to make calls to endpoint

8888 committed:
add logout functionality to remove token and clear localstorage

8888 committed:
reorganize server file structure

iovis commented on schema.sql:
```diff
@@ -0,0 +1,63 @@
+-- Name: user_auth; Type: SCHEMA; Schema: -; Owner: postgres
+--
+
+CREATE SCHEMA user_auth;
```
Nope, this should be in the public schema. Schemas are like namespaces inside a database and you only have one table for now haha.

Tbh I've never seen multiple schemas being used by anybody.

iovis commented on schema.sql:
```diff
@@ -0,0 +1,63 @@
+CREATE SCHEMA user_auth;
+
+
+ALTER SCHEMA user_auth OWNER TO postgres;
```
Next time use `--no-owner` because you don't know what user everybody is going to use. Also, you shouldn't use `postgres` because it's an administrator role.

iovis commented on schema.sql:
```diff
@@ -0,0 +1,63 @@
+    salt text NOT NULL,
+    token text NOT NULL,
+    username text NOT NULL,
+    pass_hash text NOT NULL
```
None of this is supposed to be unlimited data, so you should use a `varchar(n)` type —being `n` a max length integer—. Performance varies for lookups between varchar and text and storage is also optimized.

iovis commented on schema.sql:
```diff
@@ -0,0 +1,63 @@
+-- Name: auth_id_seq; Type: SEQUENCE; Schema: user_auth; Owner: postgres
+--
+
+CREATE SEQUENCE user_auth.auth_id_seq
```
You shouldn't need to create your own sequence unless it has something special on it

8888 replied:
Is there just a stock incrementing sequence to use in cases like this? Is that making id `serial` type will do?

iovis replied:
Yup, that would be correct

iovis commented on schema.sql:
```diff
@@ -0,0 +1,63 @@
+    id integer DEFAULT nextval('user_auth.auth_id_seq'::regclass) NOT NULL,
+    salt text NOT NULL,
+    token text NOT NULL,
+    username text NOT NULL,
```
You probably want this to be unique

iovis commented on schema.sql:
```diff
@@ -0,0 +1,63 @@
+-- Name: auth; Type: TABLE; Schema: user_auth; Owner: postgres
+--
+
+CREATE TABLE user_auth.auth (
```
`auth` as a table name tells me nothing. I think you should go for `users` for now

iovis commented on schema.sql:
```diff
@@ -0,0 +1,63 @@
+    token text NOT NULL,
+    username text NOT NULL,
+    pass_hash text NOT NULL
+);
```
Without having seen the rest of the code yet, I think you want some like the following:
```sql
CREATE TABLE users(
  id serial PRIMARY KEY,
  username VARCHAR (50) UNIQUE NOT NULL,
  password VARCHAR (50) NOT NULL,
  salt VARCHAR (50) NOT NULL,
  token VARCHAR (50) NOT NULL
);
```

8888 replied:
Only change is I removed token as per your comments below

iovis commented on README.md:
````diff
@@ -1 +1,55 @@
+Run a docker image with postgres. You can run without the `-v` arg and value to prevent data from persisting and not need to create the directories.
+
+```
+$ docker run --rm --name pg-docker -e POSTGRES_PASSWORD=docker -d -p 5432:5432 -v $HOME/docker/volumes/userAuth:/var/lib/postgresql/data postgres
````
This is personal preference for me and I mention it only in case it's useful to you.

I prefer using `docker-compose`, even if it's only for one service. It kinda abstracts away everything for me.

```yaml
-- docker-compose.yml
-- You can also put volumes here, but I'm not a volume guy
version: '3.1'
services:
  postgres:
    image: postgres:11.2
    environment:
      - POSTGRES_USER=$USER
    ports:
      - 5432:5432
```

Notice I didn't put a password and I used the current user ('david' for me, 'lee' for you) as the database user. This is so you can do:
```sh
postgresql://localhost/user_auth
# instead of
postgresql://postgres:docker@localhost:5432/user_auth
```

You start it with `docker-compose up -d --remove-orphans`.

For updating the db you can just:
```sh
createdb -h localhost user_auth
psql -h localhost -f schema.sql user_auth
```

Because you're already publishing those ports, no need to go through docker and its cryptic commands.

If you use `docker-compose stop` it'll stop the VM but won't delete the database, so you can `docker-compose up` again without risk of losing the db.

If you want to clean up, you just run `docker-compose down` and it'll remove all the data.

8888 replied:
This is awesome! I changed all of it but the one thing that is giving me trouble is my dump command. I originally had
```
docker exec pg-docker pg_dump -s -U postgres user_auth > schema.sql
```
but this uses `-U postgres` which doesn't exist anymore. If I remove it and run
```
docker exec user_auth_postgres_1 pg_dump -s user_auth > schema.sql
```
It tells me `role "root" does not exist`. If I try with the current user like
```
docker exec user_auth_postgres_1 pg_dump -s -U $USER user_auth > schema.sql
```
this works, but then my dump file that anyone else would restore from alters the tables to owner leecostello. I feel like that is incorrect. Am I missing a step and the `postgres` user should still exist or is there something else I should be doing?

note: the switch from `pg-docker` to `user_auth_postgres_1` is because the name of the container changed

In psql `\du` only shows one row, with `Role name` = `leecostello`

iovis replied:
I read this _after_ making comments in the review, lol. If you execute `pg_dump` through docker then your user is `root` (because you're in docker). I gave you a command on the review that I think should work. Let me know if it doesn't work or if you have trouble understanding something.

8888 replied:
This worked much better! Only change I had to make was the version of the docker image. It had mismatched versions of the database and pg_dump.

iovis commented on README.md:
````diff
@@ -1 +1,55 @@
+
+# Connect to psql
+```
+$ psql -h localhost -U postgres -d postgres
````
If you don't want to remember the flags you can use the url syntax as well

iovis commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+node_modules/
+
+# editors
+.vscode/
```
This should go into your global gitignore: https://help.github.com/en/articles/ignoring-files#create-a-global-gitignore

Local gitignores should only ignore project artifacts

iovis commented on dbInterface.js:
```diff
@@ -0,0 +1,86 @@
+
+const { Pool } = require('pg');
+
+class DbInterface {
```
You probably have started noticing this file is not going to scale. It has too many responsibilities and it'll grow to be a God class.

In nodejs (and in general) you should aim to have the least responsibilities as possible.

I recommend the following:
1. Make this an abstraction of the db
2. Follow the Repository Pattern or write services

### 1. Make this an abstraction of the db 
Node people like the convenience of `index.js` files. Let's try it to see how it works.

Create a `app/db/index.js` file with the following:
```js
// app/db/index.js
const { Pool } = require('pg');

const pool = new Pool({ connectionString: 'postgresql://localhost/user_auth' });

// publishing my interface and keeping stuff private
module.exports = {
  query: (text, params, callback) => pool.query(text, params, callback)
}
```

With this you can now go to your node console and use it:
```js
> db = require('./app/db')
> db.query('select 1', (err, res) => console.log(res.rows))
[ { '?column?': 1 } ]
```

You have successfully abstracted both the connection details and the db interface. This means that if you change the database or the npm package you just have to change this file.

I'm going to go a bit further, I'm not a fan of the current `query` implementation for the PG package. But I saw you can pass an object [like so](https://node-postgres.com/features/queries#query-config-object):
```js
const query = {
  text: 'INSERT INTO users(name, email) VALUES($1, $2)',
  values: ['brianc', 'brian.m.carlson@gmail.com'],
}
```

This means we could probably use a [tagged template](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_templates), which I'm clearly a fan of:
```js
const sql = (strings, ...values) => {
  const text = strings.reduce((query, current, idx) => `${query}$${idx}${current}`);
  return { text, values };
}

// with this you can just
username = 'david.marchante@rubiconmd.com';
sql`select * from users where username = ${username};`

// These would be the entry args (you can see we'd just pass through 'values')
// { strings: [ 'select * from users where username = ', ';' ],
//   values:  [ 'david.marchante@rubiconmd.com' ] }
//
// And this the reconstructed query
// 'select * from users where username = $1;'

// So now you could:
const user = await db.query(sql`select * from users where username = ${username};`);
```

### 2. Follow the Repository Pattern or write services
I think you should write **services**, but what you were trying to do was basically a repository pattern.

If you want to do a repository pattern, follow the following structure:
```
repositories/
  userRepository.js   
  sessionRepository.js
  // and so on
```

```js
class userRepository {  // could probably extend a common interface
  constructor(db) {}  // always pass an instance of db to invert the dependency

  find(id): User {}
  create(user: User): User {}
  update
  remove
  // and so on
}
```

I don't like the repository pattern very much because it's a responsibility magnet. Instead I like to use services and query objects:

```js
services/
  createUser.js
queries/
  latestOffers.js
```

### Extra:
I see you made a `rowExists` method, but it's a bit expensive BE side right now. 
When you query for the existence of something you don't need to bring any data with you, just a boolean. I recommend using `SELECT EXISTS(subquery);`:
```js
module.exports = {
  query: (text, params, callback)  => pool.query(text, params, callback),
  exists: async (sqlObject) => {
    sqlObject.text = `select exists(${sqlObject.text});`;
    const result = await pool.query(sqlObject);
    return result.rows[0].exists;
  }
};

// in other file
await db.exists(sql`select * from user_auth.auth where username = ${'david.marchante@rubiconmd.com'}`);
// true
```

8888 replied:
I only had a little bit of time and got through understanding and coding up section 1. This is such a difference though, the db abstraction and tag are fantastic. Did you just then think of making that sql tag or is this a normal pattern? It's super useful!

Also, for file structure I nested the tag in the db dir, does that feel right? Or is it super granular and that should even be `app/sql/index.js`?

iovis replied:
It's the first time I go this low level with node. The tagged template just felt natural after refactoring the db object.

Refactoring the db object to expose just a minimal interface is indeed a common pattern though. It's called _Dependency Inversion_ (the one Dependency Injection is based on) and it basically means that instead of depending on low level modules you depend on abstractions (so you _invert_ the dependency). This means you can swap objects (and mock them) even if you change libraries.

As for where to put the `sql` function, in `db/` is fine. I wouldn't go too crazy with granularity.

8888 replied:
Commits for these comments are pushed up. I went with the services and queries approach. Let me know if that looks like what you were thinking. I've now fully removed that original DB god class! This `sql` tag function is so nice...

iovis replied:
Hahaha, I'm glad you liked it!

iovis replied:
I didn't talk, but maybe I should've, about the Active Record pattern —the one we use in Rails—.

The Active Record pattern is a favorite for a lot of ORMs and it makes the model responsible of its own data layer abstraction as well. So you have things like (Rails' ActiveRecord ORM):
```rb
user = User.find_by(name: 'Lee')
user.name = 'Lee 🧛🏻‍♂️'
user.save
```

Separating the business logic from the data layer would be the Data Mapper pattern, which may not look as pretty, but is more easily optimizable.

There's no silver bullet and in the end what matters is that you know when to use what.

iovis commented on auth.js:
```diff
@@ -0,0 +1,41 @@
+const crypto = require('crypto');
+const DbInterface = require('./dbInterface.js');
+
+class Auth {
```
I'll come back and review this later, this is a responsibility magnet that you'll veeeery likely need to split.
For now I'm focusing on architecture

8888 replied:
We'll destroy this monster soon!

iovis commented on auth.js:
```diff
@@ -36,6 +36,27 @@ class Auth {
+      const hash = this.hashPassword(password, saltResult.salt);
+      if (await this.dbInterface.userAndPassMatches(username, hash)) {
+        const token = this.generateRandomValue();
+        if (await this.dbInterface.setToken(username, token)) {
```
It looks like you're doing a lot of database roundtrips here. Maybe you want to just fetch the _User_ (hint, hint)

iovis commented on README.md:
```diff
@@ -47,7 +47,7 @@ CREATE SEQUENCE user_auth.auth_id_seq
   id integer DEFAULT nextval('user_auth.auth_id_seq'::regclass) NOT NULL,
   salt text NOT NULL,
-  token text NOT NULL,
+  token text,
```
I'm guessing this `token` is the session identifier. Not a fan that it lives inside what it should be the `users` table.
You should consider creating a `sessions` table with a Foreign Key to this table.

sessions
id|token|user_id|created_at

As a general rule of thumb, think of responsibilities and ownership. 
Is the _User_ responsibility to set their own password? Yes, they're the ones that have to remember it.
Is the _User_ responsibility to manage their own session? No, it's ours and it should be transparent to them. This means it's likely that there's other behavior attached to this that we're not seeing yet (as a hint, I'll ask you to expire the session at some point, but not yet).

8888 replied:
great insight thanks, I added the `sessions` table

iovis commented on auth.js:
```diff
@@ -36,6 +36,27 @@ class Auth {
     return response;
   }
+
+  async loginUser(username, password) {
```
Not going to lie, this looks a lot like a _Controller_

iovis commented on dbInterface.js:
```diff
@@ -83,6 +85,29 @@ class DbInterface {
+      WHERE username=$1;
+    `;
+    const values = [username];
+    try {
```
You're going to be so happy once I teach you how to do a catch-all for all of these, but your architecture doesn't support it yet :D

iovis commented on server.js:
```diff
@@ -0,0 +1,29 @@
+'use strict';
```
Btw, no one uses `use strict` in node

iovis replied:
Also, let's rename this file to `app.js` to adhere to conventions

8888 replied:
Curious, why do we not do `use strict` in node? Doesn't node use v8?

iovis replied:
I honestly have no idea why people in node don't use it. It does work, I just don't know if people just rely on linters.

iovis commented on server.js:
```diff
@@ -0,0 +1,29 @@
+const Auth = require('./auth.js');
+
+const app = express();
+app.use(bodyParser.json());
```
Welcome to your first middleware 🎉 .

Middleware is basically stuff that happens from the request hitting the app to the request actually hitting your own code.

This usually entails serializing objects, routing, logging... But you can `use` _:D_ it for whatever you want. I, for instance, like to use it for decoupling validations and auth checks.

Here, as you can see, it's just serializing your request to JSON.

Let's do our first one.

# Responsibilites

It looks like my favorite word. You're probably sick of me saying it.

You're already putting too much stuff here:
- Instantiates the server
- Puts the middleware
- Runs the server
- Instantiates `Auth`
- Does Routing
- Has controllers
- Does views

And it's only 29 lines long and just has one endpoint. This is not going to scale.

## Extracting the Routes

Create a `routes` folder. It seems like you're namespacing this one to `user` (you should use the plural for objects, so I'll go for `users` from now on).

Perfect, let's create a namespaced routing system for users:
```js
// routes/users.js
const express = require('express');
const router = express.Router();
const Auth = require('../auth');

const auth = new Auth();

// Notice I don't need the `/users` namespace
router.post('/register', async (req, res) => {
  const credentials = req.body;
  if (credentials.username && credentials.password) {
    const result = await auth.registerUser(credentials.username, credentials.password);
    if (result.success) {
      res.status(201).json({status: 201, user: credentials.username});
    } else {
      res.status(409).json({status: 409, error: result.message});
    }
  } else {
    res.status(400).json({status: 400, error: 'Data is invalid'});
  }
});

module.exports = router;
```

Now add your routing as middleware in `app.js`:
```js
// app.js
const express = require('express');
const bodyParser = require('body-parser');
const usersRouter = require('./routes/users');

const app = express();
app.use(bodyParser.json());
app.use('/users', usersRouter);  // <-- :D

const server = app.listen(process.env.PORT || 8080, () => {
  const port = server.address().port;
  console.log(`App now running on port ${port}`);
});
```

Ok, a lot less responsibilities for `app.js`. Now it only has to set the middleware and run the server.

But our Routing now has to do controller and view stuff. That's not cool...

iovis replied:
_The answer to that will come in the next episode_

iovis replied:
Ok, next episode is _ON_.

# Controllers

In today's episode we're going to start creating _Controllers_. As you may know, controllers orchestrate different parts of the application together, with one caveat: 
_controllers are not necessarily supposed to know what they're doing_.

I see this a lot, even in our application. People think that because controllers are the ones in charge of orchestrating stuff they should know the nitty-gritty details of everything. That results in a controller being 50 lines long —when it should be 5-10 lines long—.

Controllers are like managers. They tell people what to do, but don't necessarily know how to do it, they just know who to ask and how to orchestrate them. Tim knows it has to tell _you_ to change the color of a button, it doesn't mean he _knows_ how to do it (or even should).

In a well written application, those _responsibilities_ (my favorite word) go to _Service Objects_.

There's a lot of ways to do SOs, but I like them to have just one responsibility (or just one reason to change), following the _Single Responsibility Principle_ (from now on **SRP**). 
I also like them to have the same _interface_ (following the _Liskov Substitution Principle_) which would be having just one public method that it's always the same (in Ruby we use `.call`, because it's used heavily by other things in Ruby).
This also means you're not forcing your _clients_ to rely on methods they _don't need_ (thus following the _Interface Segregation Principle_).
If you add things like what you did with the `db` object, you would be using SOs to _invert the dependency_ and rely on abstractions instead of low level details (following the _Dependency Inversion Principle_) then you can basically see we're aiming for a _[SOLID](https://en.wikipedia.org/wiki/SOLID)_ design :D.

Okay, let's refactor. We have the following right now:

```js
// routes/users.js
router.post('/register', async (req, res) => {
  const credentials = req.body;
  if (credentials.username && credentials.password) {
    const result = await auth.registerUser(credentials.username, credentials.password);
    if (result.success) {
      res.status(201).json({status: 201, user: credentials.username});
    } else {
      res.status(409).json({status: 409, error: result.message});
    }
  } else {
    res.status(400).json({status: 400, error: 'Data is invalid'});
  }
});
```

A routing module should not care about all of this, let's extract this into a controller:
```js
// controllers/users.js
const Auth = require('../auth');
const auth = new Auth();

exports.register = async (req, res) => {
  const credentials = req.body;

  if (credentials.username && credentials.password) {
    const result = await auth.registerUser(
      credentials.username,
      credentials.password
    );
    if (result.success) {
      res.status(201).json({ status: 201, user: credentials.username });
    } else {
      res.status(409).json({ status: 409, error: result.message });
    }
  } else {
    res.status(400).json({ status: 400, error: 'Data is invalid' });
  }
};
```
And now our router looks like this:
```js
// routes/users.js
const express = require('express');
const router = express.Router();
const usersController = require('../controllers/users');

router.post('/register', usersController.register);

module.exports = router;
```

_HOT DAMN_

Now the router only cares about routing. My parents were right: I'm always finding ways to avoid responsibilities.

But now I look at the controller and it still feels wrong. I should be able to look at it and just _know_ what it does, but I get lost in a sea of `if` statements... 

This controller knows _too much_... We'll have to deal with it 🔫

But that will be on the next episode:
We'll delve on Service Objects, Models (even _top_ models), some more middleware and we'll probably break your _Auth_ object apart :D

8888 replied:
app and the users route looks so good now! This entire section was fantastic! Tons of good reading and I see what you mean about the users controller I have now, I'm excited to keep refactoring these down.

iovis commented on server.js:
```diff
@@ -0,0 +1,29 @@
+    if (result.success) {
+      res.status(201).json({status: 201, user: credentials.username});
+    } else {
+      res.status(409).json({status: 409, error: result.message});
```
Nope. I'd use either `422` or, in rare cases, `418`.

iovis replied:
Also, the status is not meant to be sent in the view. That's just an HTTP header.

8888 replied:
418 is the obvious choice

iovis replied:
haha, I knew you'd like it

iovis commented on server.js:
```diff
@@ -0,0 +1,29 @@
+const auth = new Auth();
+
+app.post('/user/register', async (req, res) => {
+  const credentials = req.body;
```
I'd go for `const { username, password } = req.body` but I'm a psychopath

8888 replied:
this is one of my favorite lines from these review comments

iovis replied:
Shit, then I don't know if I should tell you that I think that this would work:
```js
async ({ body: { username, password } }, res) => {
  console.log(username);
  console.log(password);
});
```

iovis commented:
Btw, I recommend adding `nodemon`:
```
npm i -D nodemon
```

`package.json`
```json
"scripts": {
    "watch": "nodemon app.js",
}
```

iovis commented on index.html:
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
```
I you're serving static files you should let Express do it for you:

1. Create a `public/` folder
2. Move every static asset there
3. Move everything in your `app/*` folder to the root (we'll reorganize later)
4. Rename your `server.js` to `app.js` to follow conventions
5. Add the following line in your `app.js` (as the first middleware!)
```js
app.use(express.static(path.join(__dirname, 'public')));  // you'll need 'path' from require('path');
```

8888 committed:
replace single table and create users table

this handles only the user responsibilities. the token was removed and will be added to another table next

8888 committed:
created sessions table

8888 committed:
move editors from local gitignore to global

8888 committed:
created docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to changecreated docker-compose yml

updated readme to reflect the changes. schema.sql changed based on the new database formatting. The owner of the tables may still need to change

iovis commented on README.md:
```diff
@@ -47,4 +47,10 @@ CREATE TABLE users(
+  id serial PRIMARY KEY,
+  token VARCHAR (50) NOT NULL,
+  user_id INTEGER REFERENCES users(id),
+  created_at TIMESTAMPTZ NOT NULL
```
You can put a `default now()` if you want so you don't have to set it on creation

iovis commented on README.md:
```diff
@@ -1,17 +1,13 @@
 # Setup
-If you do not have the docker postgres image, it can be pulled from the docker container repo.
+Run a docker image with postgres.
```
I'm a bit scared, Lee.
![Screenshot 2019-04-09 at 13 26 53](https://user-images.githubusercontent.com/6660202/55796838-2e1ca180-5acb-11e9-8be8-a865c4f828af.png)
![download](https://user-images.githubusercontent.com/6660202/55796841-2f4dce80-5acb-11e9-9b3b-43a29e735905.jpg)

8888 replied:
This is very concerning as I have no recollection of this.

iovis commented on README.md:
````diff
@@ -20,27 +16,29 @@ $ npm run serve
 # Create a dump of the schema
 ```
-$ docker exec pg-docker pg_dump -s -U postgres user_auth > schema.sql
+$ docker exec user_auth_postgres_1 pg_dump -s -U $USER user_auth > schema.sql
````
You're exposing the ports, you don't need docker commands, and you don't need the user anymore:
```sh
pg_dump -s --clean --no-owner -h localhost user_auth > schema.sql
```

iovis commented on README.md:
````diff
@@ -20,27 +16,29 @@ $ npm run serve
 # Connect to psql
 ```
-$ psql -h localhost -U postgres -d postgres
+$ psql -h localhost -d postgres
````
`postgres` => `user_auth`

iovis commented on schema.sql:
```diff
@@ -20,7 +20,7 @@ SET default_tablespace = '';
 
 --
--- Name: sessions; Type: TABLE; Schema: public; Owner: postgres
+-- Name: sessions; Type: TABLE; Schema: public; Owner: leecostello
```
This is like marking your territory but for databases

8888 replied:
🐕

8888 committed:
update pg_dump command

the docker image version needed to be downgraded to make this work. The session timestamp was also updated to have a default value with this dump

8888 committed:
update connection to use url syntax

iovis commented on schema.sql:
```diff
@@ -0,0 +1,182 @@
+--
```
Looking much better

8888 committed:
npm i -D nodemon

in package.josn scripts, replaces serve with watch to use nodemon

8888 committed:
rename server.js to app.js

8888 committed:
create database abstraction and sql tag

iovis commented on app/db/sql.js:
```diff
@@ -0,0 +1,6 @@
+  return { text, values };
+};
+
+module.exports = { sql };
```
if you're only exporting the function then you don't need the object. 
```js
module.exports = sql;
```

Otherwise instead of:
```js
const sql = require('./db/sql');
sql`select 1`;
```
You'll be doing:
```js
const sql = require('./db/sql');
sql.sql`select 1`;

// or
const { sql } = require('./db/sql');
sql`select 1`;
```

8888 committed:
remove brackets for a cleaner import syntax

8888 committed:
create exists method on db

8888 committed:
move usernameExists to a query

8888 committed:
move userAndPassMatches to a quewry

8888 committed:
move userIsAuthorized to a query

this will need an update later because tokens will change. Right now as long as this user has ever had a session token they are authorized

8888 committed:
remove code from db interface that has been pulled out

the rest of the methods in this class are being removed as well. This is more of a todo list at this point

8888 committed:
create setToken function

8888 committed:
create addUser function

8888 committed:
create fetchSalt function

this also removed my monster dbInterface file! wooooo

8888 committed:
change length of salt and password hash

8888 committed:
update auth to use the db changes

this will all need refactoring, but this was just to get it working to first start working on the app middleware

8888 committed:
move register to users route middleware

8888 committed:
move login to users route

8888 committed:
moved confirmToken to users route

8888 committed:
move logout to users route

8888 committed:
remove unused dependency from app

8888 committed:
move register into the newly made users controller

8888 committed:
move login to users controller

8888 committed:
move confirmToken to users controller

8888 committed:
move logout to users controller

8888 committed:
remove extra db roundtrip from login

consolidated fetching the salt and matching the hashed passwords into one query

8888 committed:
use express to serve static content

this rearranged the file structure a bit

iovis commented on app/queries/usernameExists.js:
```diff
@@ -0,0 +1,6 @@
+const sql = require('../db/sql.js');
+
+module.exports = async (username) => {
+  return await db.exists(sql`select * from users where username = ${username}`);
```
I think it's better that you just return the promise and let the code that uses this decide whether to use `await`. You might want to use it differently

iovis commented on app/queries/userIsAuthorized.js:
```diff
@@ -0,0 +1,12 @@
+    select *
+    from users
+    join sessions on users.id = sessions.user_id
+    where users.username = ${username}
```
Damn, even GitHub likes the tagged template. It does proper sql syntax highlighting and everything, wtf

iovis commented on app/services/setToken.js:
```diff
@@ -0,0 +1,11 @@
+    values (${token}, current_timestamp,
+      (select id from users where username = ${username})
+    )
+  `);
```
also, one other thing aside from the `await` thing, you might want to leverage JS's implicit `return` when you only do one thing:
```js
module.exports = (username, token) => (  // <-- Notice I used () for having a multi-line one-liner
  db.query(sql`
    insert into sessions(token, created_at, user_id)
    values (${token}, current_timestamp,
      (select id from users where username = ${username})
    )
  `);
);
```

iovis commented on app/services/addUser.js:
```diff
@@ -0,0 +1,9 @@
+  return await db.query(sql`
+    insert into users (username, password, salt)
+    values (${username}, ${password}, ${salt});
+  `);
```
When creating an object is customary to return said object with the ID. But let me review the rest of the project before doing that

iovis commented on app/queries/fetchSalt.js:
```diff
@@ -0,0 +1,7 @@
+
+module.exports = async (username) => {
+  const result = await db.query(sql`select salt from users where username = ${username}`);
+  return result.rows[0].salt;
```
If you only want one row then prefer using `limit 1` in the query

iovis replied:
Nvm, I think I misinterpreted your intention here haha

iovis commented:
Damn son, you were _busy_ refactoring haha.

This is starting to look great, I'll try to make a deep dive today to see what will be the next steps

iovis commented on app/routes/users.js:
```diff
@@ -6,20 +6,7 @@ const usersController = require('../controllers/users');
-    res.status(401).json({ error: 'User is not authenticated!' });
-  }
-});
+router.post('/confirmToken', usersController.confirmToken);
```
BE people don't like camelCase for URLs, it's kinda frowned upon and a common tell from FE people designing APIs haha.

It's recommended to use hyphens because it's a non-word character.

iovis replied:
One example of Google recommending doing it like that: https://support.google.com/webmasters/answer/76329?hl=en

iovis commented:
So I sometimes like to play around with my code in the console and I haven't found a good way in Node world to do this, so I came up with a naive implementation of what I mean. Maybe you find it useful:

`./console`
```js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

const requireDir = (dir) => (
  fs.readdirSync(dir).reduce((obj, filename) => {
    const file = path.parse(filename);

    return file.ext === '.js'
      ? { ...obj, [file.name]: require(path.join(__dirname, dir, filename)) }
      : obj;
  }, {})
);

const repl = require('repl');
const replServer = repl.start();
const context = replServer.context;

// Adding stuff
context.db = require('./db');
context.sql = require('./db/sql');
context.services = requireDir('services');
context.controllers = requireDir('controllers');
context.queries = requireDir('queries');
```

Now you can just go to the terminal and do:
```sh
$ ./console
> user = queries.fetchUser('david.marchante@rubiconmd.com')
```

Which I think is pretty neat

iovis commented:
I actually found it so useful that I made it a bit better and published it to npm:
https://www.npmjs.com/package/requiem-tree

Now you can `npm install requiem-tree --save-dev` and `console` would look like:
```js
#!/usr/bin/env node

const requiem = require('requiem-tree');
const repl = require('repl');
const replServer = repl.start();
const context = replServer.context;

// Adding stuff
context.db = requiem('db');
context.services = requiem('services');
context.controllers = requiem('controllers');
context.queries = requiem('queries');
```

iovis commented on auth.js:
```diff
@@ -0,0 +1,83 @@
+const setToken = require('./services/setToken');
+const addUser = require('./services/addUser');
+
+const generateRandomValue = (bytes = 32) => {
```
Now you should split this file into all the services it comprises and require those only in the controllers/services you need them

iovis commented on controllers/users.js:
```diff
@@ -0,0 +1,53 @@
+
+exports.register = async (req, res) => {
+  const { username, password } = req.body;
+  if (username && password) {
```
Following the idea of Controllers just telling people what to do, we should follow the **_Tell, Don't Ask_** principle.

It just means that you ask whatever thing is responsible of doing it to do it, and it's the other object's responsibility to validate it:
```js
exports.register = async (req, res) => {
  const { username, password } = req.body;
  
  try {
    const user = await RegisterUser.call({ username, password });
    res.status(201).json({ user });
  catch (error) {
    res.status(422).json({ error });
  }
};
